tests/Microsoft.Identity.Test.Unit.net45/Integration/DeviceCodeFlowIntegrationTest.cs

```diff
@@ -0,0 +1,169 @@
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using OpenQA.Selenium;
+using OpenQA.Selenium.Chrome;
+using OpenQA.Selenium.Support.UI;
+
+namespace Microsoft.Identity.Test.Unit.Integration
+{
+    // Important: do not install a NuGet package with the Chrome driver as it is a security risk.
+    // Instead, install the Chrome driver on the test machine
+
+    // Note: these tests require permission to a KeyVault Microsoft account; 
+    // Please ignore them if you are not a Microsoft FTE, they will run as part of the CI build
+    [TestClass]
+    public class DeviceCodeFlow
+    {
+        private static readonly string[] Scopes = { "User.Read" };
+
+        public TestContext TestContext { get; set; }
+
+        [ClassInitialize]
+        public static void ClassInitialize(TestContext context)
+        {
+            ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls12;
+        }
+
+        [TestInitialize]
+        public void TestInitialize()
+        {
+            TestCommon.ResetState();
+        }
+
+
+        [TestMethod]
+        [Timeout(60 * 1000)] // 1 min timeout
+        public async Task DeviceCodeFlowTestAsync()
+        {
+            Debug.WriteLine("Fetching user from lab");
+            LabResponse labResponse = LabUserHelper.GetDefaultUser();
+
+            Debug.WriteLine("Calling AcquireTokenWithDeviceCodeAsync");
+            PublicClientApplication pca = new PublicClientApplication(labResponse.AppId);
+            var result = await pca.AcquireTokenWithDeviceCodeAsync(Scopes, deviceCodeResult =>
+            {
+                RunAutomatedDeviceCodeFlow(deviceCodeResult, labResponse.User);
+
+                return Task.FromResult(0);
+            }).ConfigureAwait(false);
+
+            Assert.IsNotNull(result);
+            Assert.IsTrue(!string.IsNullOrEmpty(result.AccessToken));
+        }
+
+        private void RunAutomatedDeviceCodeFlow(DeviceCodeResult deviceCodeResult, LabUser user)
+        {
+            IWebDriver driver = null;
+            try
+            {
+                driver = InitDriver();
+
+                Debug.WriteLine("Navigating and enterign the code");
+
+                driver.Navigate().GoToUrl(deviceCodeResult.VerificationUrl);
+                driver.FindElement(By.Id("code")).SendKeys(deviceCodeResult.UserCode);
+
+                IWebElement continueBtn = WaitForElementToBeEnabled(driver, By.Id("continueBtn"));
+                continueBtn?.Click();
+
+                Debug.WriteLine("Logging in");
+                PerformLogin(driver, user);
+
+            }
+            catch (Exception ex)
+            {
+                Debug.WriteLine("Browser automation failed " + ex);
+#if DESKTOP // Can't attach a file on netcore because mstest doesn't support it
+                Screenshot ss = ((ITakesScreenshot)driver).GetScreenshot();
+                string failurePicturePath = Path.Combine(TestContext.ResultsDirectory, "failure.png");
```

Comment:
>failure [](start = 88, length = 7)

should probably have a bit of randomness to the name or a time in case there will end up being more than one screenshot. could also be the test name?

Reply:
Yes, good idea.